shipwright/source_control.py

```diff
@@ -52,6 +55,40 @@ def _hexsha(ref):
         return ref.hexsha[:12]
 
 
+def _hash_blob(blob):
+    if blob.hexsha != blob.NULL_HEX_SHA:
+        return blob.hexsha
+    else:
+        return blob.repo.git.hash_object(blob.abspath)
+
+
+def _dirty_suffix(repo, base_paths=['.']):
+    def abspath(path):
+        return os.path.abspath(os.path.join(repo.working_dir, path))
+
+    def hash_blobs(blobs):
+        return [(b.path, _hash_blob(b)) for b in blobs if b]
+
+    def in_paths(path):
```

Comment:
this doesn't need to be a closure, move it to _in_paths(repo_wd, base_path, path) at module scope